migrations/utils.py

```diff
@@ -0,0 +1,167 @@
+            return self._data == other.data
+        return False
+
+    def __hash__(self):
+        """Hash the node's data."""
+        return hash(self._data)
+
+
+class LinkedList:
+    def __init__(self):
+        """Initialize a linked list."""        
+        self._head = None
+
+    def find(self, data) -> Node:
+        """Find the first node containing the specified data."""
+        current_node = self._head
+        while current_node:
+            if current_node.data == data:
+                return current_node
+            current_node = current_node.prev_node
+        return None
+
+    def next_node(self, current_node_data) -> Node:
+        """Retrieve node following the specified one."""
+        next_node = self._head
+        while next_node and next_node.prev_node:
+            if next_node.prev_node.data == current_node_data:
+                return next_node
+            next_node = next_node.prev_node
+        return None
+
+    def next(self, current_node_data) -> object:
+        """Retrieve data following the specified one."""
+        node = self.next_node(current_node_data)
+        if node:
+            return node.data
+        else:
+            return None
+
+    def previous_node(self, current_node_data) -> Node:
+        """Retrieve node before the specified one."""
+        current_node = self.find(current_node_data)
+        if current_node and current_node.prev_node:
+            return current_node.prev_node
+        return None
+    
+    def previous(self, current_node_data) -> object:
+        """Retrieve data before the specified one."""
+        node = self.previous_node(current_node_data)
+        if node:
+            return node.data
+        else:
+            return None
+    
+    def get_sublist(self, first_node_data: str, last_node_data: str = None) -> list:
+        """Return a list consistings of nodes between the specified boundaries.
+        Inclusive of last endpoint, not of first."""
+        unapplied_migrations = []
+        if last_node_data is None:
+            last_node = self._head
+        else:
+            last_node = self.find(last_node_data)
+
+        # Iterate over migrations starting from top
+        while last_node and last_node.data != first_node_data:
+            unapplied_migrations.append(last_node.data)
+            last_node = last_node.prev_node
+
+        # Reverse to account for the order
+        unapplied_migrations.reverse()
+
+        return unapplied_migrations
+
+    def add(self, data):
+        """Add new node after the head. Raises an error if node with such migration already exists."""
+        new_node = self.find(data)
+        if new_node:
+            raise ValueError("adding a duplicate item")
+        else:
+            new_node = Node(data)
+
+            if self._head is None:
+                self._head = new_node
+            else:
+                current_node = self._head
+                while current_node.next_node:
```

Comment:
aren't we adding a single node after `self._head` on every call here?  the `while` seems unnecessary, wouldn't it be exceptional to find `self._head.next_node != None` ?